File: CS/MyXtraGrid/MyPrintInfo.cs
using System;
using DevExpress.XtraGrid.Views.Printing;
using System.Drawing;
using DevExpress.XtraPrinting;
using DevExpress.Utils;
using DevExpress.Utils.Drawing;
using DevExpress.XtraPrinting.NativeBricks;
using System.Windows.Forms;

namespace MyXtraGrid {
    public class MyPrintInfo : GridViewPrintInfo {

        public MyPrintInfo(PrintInfoArgs args)
            : base(args) {
        }

        private void MakeInflate(ref ITextBrick itb, float offsetByHorizontalAxis, float offsetByVerticalAxis)
        {
            itb.Rect = new RectangleF(itb.Rect.Location, new SizeF(itb.Rect.Width - offsetByHorizontalAxis, itb.Rect.Height - offsetByVerticalAxis));
        }

        public override void PrintHeader(IBrickGraphics graph)
        {
            if (!View.OptionsPrint.PrintHeader) return;

            Point indent = new Point(Indent, HeaderY);
            Rectangle r = Rectangle.Empty;
            bool usePrintStyles = View.OptionsPrint.UsePrintStyles;
            SetDefaultBrickStyle(graph, Bricks["HeaderPanel"]);

            foreach (PrintColumnInfo col in Columns)
            {
                if (!usePrintStyles)
                {
                    AppearanceObject temp = new AppearanceObject();
                    AppearanceHelper.Combine(temp, new AppearanceObject[] { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel });
                    SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1));
                }
                r = col.Bounds;
                r.Offset(indent);
                string caption = col.Column.GetTextCaption();
                if (!ColumnsInfo[Columns.IndexOf(col)].Column.OptionsColumn.ShowCaption) caption = string.Empty;

                    DevExpress.XtraGrid.Drawing.GridColumnInfoArgs columnsInfo = ColumnsInfo[Columns.IndexOf(col)];
                    DrawElementInfoCollection innerElements = columnsInfo.InnerElements;
                    DrawElementInfo columnImageInfo = null;
                    GlyphElementInfoArgs elementInfo = null;
                    this.ViewViewInfo.Painter.ElementsPainter.Column.CalcObjectBounds(columnsInfo);
                    for (int i = 0; i < innerElements.Count; i++)
                    {
                        if (innerElements[i].ElementInfo is DevExpress.Utils.Drawing.GlyphElementInfoArgs)
                        {
                            columnImageInfo = innerElements[i];
                            elementInfo = columnImageInfo.ElementInfo as GlyphElementInfoArgs;
                            break;
                        }
                    }

                    ITextBrick tBrick = null;
                    if (elementInfo.Glyph == null && elementInfo.ImageIndex < 0)
                    {
                        tBrick = DrawTextBrick(graph, caption, r);

                    }
                    else
                    {
                        IPanelBrick panelBrick = new XETextPanelBrick(graph.DefaultBrickStyle);
                        float offsetForBorder = panelBrick.BorderWidth;

                        if (columnImageInfo.Alignment != StringAlignment.Center)
                        {
                            tBrick = new TextBrick();
                            tBrick.Rect = new RectangleF(offsetForBorder, offsetForBorder, r.Width, r.Height);
                            tBrick.Text = caption;
                            tBrick.Style = graph.DefaultBrickStyle;
                            tBrick.Sides = BorderSide.None;
                            panelBrick.Bricks.Add(tBrick);
                        }
                        
                        panelBrick.Value = caption;
                        Rectangle columnRect = r;
                        Rectangle imageRect = new Rectangle(new Point((int)offsetForBorder, (int)offsetForBorder), elementInfo.GlyphSize);
                        imageRect.Y = r.Y + columnImageInfo.ElementInterval;

                        switch (columnImageInfo.Alignment)
                        {
                            case StringAlignment.Near:
                                tBrick.Rect = new RectangleF(new PointF(imageRect.Location.X + imageRect.Size.Width, tBrick.Rect.Y),tBrick.Rect.Size);
                                MakeInflate(ref tBrick, 2 * offsetForBorder, 2 * offsetForBorder);
                                break;
                            case StringAlignment.Center:
                                {
                                    imageRect.X += (r.Width - imageRect.Width) / 2;
                                }
                                break;
                            case StringAlignment.Far:
                                tBrick.Rect = new RectangleF(tBrick.Rect.Location, new SizeF(tBrick.Rect.Width - imageRect.Width, tBrick.Rect.Height));
                                MakeInflate(ref tBrick, 2 * offsetForBorder, 2 * offsetForBorder);
                                imageRect.X = (int)tBrick.Rect.Right;
                                break;
                        }


                        ImageBrick iBrick = GetImageBrick(imageRect, columnImageInfo);
                        if (iBrick != null)
                            panelBrick.Bricks.Add(iBrick);
                   
                        graph.DrawBrick(panelBrick, columnRect);
                    }
                    

                    if (tBrick == null) continue;

                    if (AppearancePrint.HeaderPanel.TextOptions.WordWrap == WordWrap.NoWrap && View.OptionsPrint.UsePrintStyles)
                    {
                        using (Graphics g = this.View.GridControl.CreateGraphics())
                        {
                            SizeF s = g.MeasureString(tBrick.Text, tBrick.Font, 1000, tBrick.StringFormat.Value);
                            if (s.Width + 5 >= r.Width)
                            {
                                tBrick.Text = "";
                                tBrick.TextValue = "";
                            }
                        }
                    }
            }
        }


        private ImageBrick GetImageBrick(Rectangle rect, DrawElementInfo element)
        {
            ImageBrick ib = new ImageBrick(BorderSide.None, 0, AppearancePrint.HeaderPanel.BorderColor, AppearancePrint.HeaderPanel.BackColor);
            ib.Rect = rect;
            GlyphElementInfoArgs glyphInfo = element.ElementInfo as GlyphElementInfoArgs;

            if (glyphInfo.Glyph != null)
            {
                ib.Image = glyphInfo.Glyph;
            }
            else
            {
                if (glyphInfo.ImageIndex < 0) return null;

                if (glyphInfo.ImageList is ImageCollection)
                {
                    ImageCollection imageList = glyphInfo.ImageList as ImageCollection;
                    ib.Image = imageList.Images[glyphInfo.ImageIndex];
                }
                else if (glyphInfo.ImageList is ImageList)
                {
                    ImageList imageList = glyphInfo.ImageList as ImageList;
                    ib.Image = imageList.Images[glyphInfo.ImageIndex];
                }
                else
                {
                    SharedImageCollection collection = glyphInfo.ImageList as SharedImageCollection;
                    ib.Image = collection.ImageSource.Images[glyphInfo.ImageIndex];
                }
            }
       
            return ib;
        }
    }
}

File: VB/MyXtraGrid/MyPrintInfo.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.XtraGrid.Views.Printing
Imports System.Drawing
Imports DevExpress.XtraPrinting
Imports DevExpress.Utils
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraPrinting.NativeBricks
Imports System.Windows.Forms

Namespace MyXtraGrid
	Public Class MyPrintInfo
		Inherits GridViewPrintInfo

		Public Sub New(ByVal args As PrintInfoArgs)
			MyBase.New(args)
		End Sub

		Private Sub MakeInflate(ByRef itb As TextBrick, ByVal offsetByHorizontalAxis As Single, ByVal offsetByVerticalAxis As Single)
			itb.Rect = New RectangleF(itb.Rect.Location, New SizeF(itb.Rect.Width - offsetByHorizontalAxis, itb.Rect.Height - offsetByVerticalAxis))
		End Sub

		Public Overrides Sub PrintHeader(ByVal graph As BrickGraphics)
			If (Not View.OptionsPrint.PrintHeader) Then
				Return
			End If

            Dim indentByBothAxes As New Point(Indent, HeaderY)
			Dim r As Rectangle = Rectangle.Empty
			Dim usePrintStyles As Boolean = View.OptionsPrint.UsePrintStyles
			SetDefaultBrickStyle(graph, Bricks("HeaderPanel"))

			For Each col As PrintColumnInfo In Columns
				If (Not usePrintStyles) Then
					Dim temp As New AppearanceObject()
					AppearanceHelper.Combine(temp, New AppearanceObject() { col.Column.AppearanceHeader, View.Appearance.HeaderPanel, AppearancePrint.HeaderPanel })
					SetDefaultBrickStyle(graph, Bricks.Create(temp, BorderSide.All, temp.BorderColor, 1))
				End If
				r = col.Bounds
                r.Offset(indentByBothAxes)
				Dim caption As String = col.Column.GetTextCaption()
				If (Not ColumnsInfo(Columns.IndexOf(col)).Column.OptionsColumn.ShowCaption) Then
					caption = String.Empty
				End If

                Dim columnInfo As DevExpress.XtraGrid.Drawing.GridColumnInfoArgs = ColumnsInfo(Columns.IndexOf(col))
                Dim innerElements As DrawElementInfoCollection = columnInfo.InnerElements
                Dim columnImageInfo As DrawElementInfo = Nothing
                Dim elementInfo As GlyphElementInfoArgs = Nothing
                Me.ViewViewInfo.Painter.ElementsPainter.Column.CalcObjectBounds(columnInfo)
                For i As Integer = 0 To innerElements.Count - 1
                    If TypeOf innerElements(i).ElementInfo Is DevExpress.Utils.Drawing.GlyphElementInfoArgs Then
                        columnImageInfo = innerElements(i)
                        elementInfo = TryCast(columnImageInfo.ElementInfo, GlyphElementInfoArgs)
                        Exit For
                    End If
                Next i

                Dim tBrick As TextBrick = Nothing
                If elementInfo.Glyph Is Nothing AndAlso elementInfo.ImageIndex < 0 Then
                    tBrick = DrawTextBrick(graph, caption, r)

                Else
                    Dim panelBrick As PanelBrick = New XETextPanelBrick(graph.DefaultBrickStyle)
                    Dim offsetForBorder As Single = panelBrick.BorderWidth

                    If columnImageInfo.Alignment <> StringAlignment.Center Then
                        tBrick = New TextBrick()
                        tBrick.Rect = New RectangleF(offsetForBorder, offsetForBorder, r.Width, r.Height)
                        tBrick.Text = caption
                        tBrick.Style = graph.DefaultBrickStyle
                        tBrick.Sides = BorderSide.None
                        panelBrick.Bricks.Add(tBrick)
                    End If

                    panelBrick.Value = caption
                    Dim columnRect As Rectangle = r
                    Dim imageRect As New Rectangle(New Point(CInt(Fix(offsetForBorder)), CInt(Fix(offsetForBorder))), elementInfo.GlyphSize)
                    imageRect.Y = r.Y + columnImageInfo.ElementInterval

                    Select Case columnImageInfo.Alignment
                        Case StringAlignment.Near
                            tBrick.Rect = New RectangleF(New PointF(imageRect.Location.X + imageRect.Size.Width, tBrick.Rect.Y), tBrick.Rect.Size)
                            MakeInflate(tBrick, 2 * offsetForBorder, 2 * offsetForBorder)
                        Case StringAlignment.Center
                            imageRect.X += CInt((r.Width - imageRect.Width) / 2)
                        Case StringAlignment.Far
                            tBrick.Rect = New RectangleF(tBrick.Rect.Location, New SizeF(tBrick.Rect.Width - imageRect.Width, tBrick.Rect.Height))
                            MakeInflate(tBrick, 2 * offsetForBorder, 2 * offsetForBorder)
                            imageRect.X = CInt(Fix(tBrick.Rect.Right))
                    End Select


                    Dim iBrick As ImageBrick = GetImageBrick(imageRect, columnImageInfo)
                    If iBrick IsNot Nothing Then
                        panelBrick.Bricks.Add(iBrick)
                    End If

                    graph.DrawBrick(panelBrick, columnRect)
                End If


					If tBrick Is Nothing Then
						Continue For
					End If

					If AppearancePrint.HeaderPanel.TextOptions.WordWrap = WordWrap.NoWrap AndAlso View.OptionsPrint.UsePrintStyles Then
						Using g As Graphics = Me.View.GridControl.CreateGraphics()
							Dim s As SizeF = g.MeasureString(tBrick.Text, tBrick.Font, 1000, tBrick.StringFormat.Value)
							If s.Width + 5 >= r.Width Then
								tBrick.Text = ""
								tBrick.TextValue = ""
							End If
						End Using
					End If
			Next col
		End Sub


		Private Function GetImageBrick(ByVal rect As Rectangle, ByVal element As DrawElementInfo) As ImageBrick
			Dim ib As New ImageBrick(BorderSide.None, 0, AppearancePrint.HeaderPanel.BorderColor, AppearancePrint.HeaderPanel.BackColor)
			ib.Rect = rect
			Dim glyphInfo As GlyphElementInfoArgs = TryCast(element.ElementInfo, GlyphElementInfoArgs)

			If glyphInfo.Glyph IsNot Nothing Then
				ib.Image = glyphInfo.Glyph
			Else
				If glyphInfo.ImageIndex < 0 Then
					Return Nothing
				End If

				If TypeOf glyphInfo.ImageList Is ImageCollection Then
					Dim imageList As ImageCollection = TryCast(glyphInfo.ImageList, ImageCollection)
					ib.Image = imageList.Images(glyphInfo.ImageIndex)
				ElseIf TypeOf glyphInfo.ImageList Is ImageList Then
					Dim imageList As ImageList = TryCast(glyphInfo.ImageList, ImageList)
					ib.Image = imageList.Images(glyphInfo.ImageIndex)
				Else
					Dim collection As SharedImageCollection = TryCast(glyphInfo.ImageList, SharedImageCollection)
					ib.Image = collection.ImageSource.Images(glyphInfo.ImageIndex)
				End If
			End If

			Return ib
		End Function
	End Class
End Namespace
